File: MarketJava/src/Client/SocketListener.java
package Client;

import java.util.NoSuchElementException;
import java.util.Scanner;

public class SocketListener implements Runnable{
    private final Scanner scanner;
    private final ClientHandler handler;


    public SocketListener(Scanner scanner, ClientHandler handler)
    {
        this.scanner = scanner;
        this.handler = handler;
    }

    @Override
    @SuppressWarnings("InfiniteLoopStatement")
    public void run()
    {
        try {
            while (true) {
                String line = scanner.nextLine();
                this.handler.processLine(line);
            }
        } catch (NoSuchElementException e)
        {
            ClientProgram.socketClosed = true;
        }
    }
}

File: MarketCSharp/Client/SocketListener.cs
using System;
using System.IO;
using System.Net.Sockets;
namespace Client
{
    /**
     * SocketListener's purpose is to read and all communication sent from the
     * server and send to the Handler to process.
     * 
     * As socket exceptions are raised in this thread, and not main thread,
     * Server rebooting initiation goes here
     */
    class SocketListener
    {
        private StreamReader reader;
        private Handler handler;

        public SocketListener(Handler handler)
        {
            this.handler = handler;
            this.reader = handler.Reader;
        }

        public void Run()
        {
            try
            {
                while (!Program.SocketClosed && !Program.applicationClosed)
                {
                    string line = reader.ReadLine();
                    if (line == null)
                        throw new Exception("Socket disconnected");
                    this.handler.ProcessLine(line);
                }
            }
            catch (Exception e)
            {
                if (!Program.applicationClosed)
                    Program.RestartServer();
            }
        }
    }
}
